tools/traffic/workers/blob_reader.go

```diff
@@ -0,0 +1,231 @@
+package workers
+
+import (
+	"context"
+	"crypto/md5"
+	"fmt"
+	"github.com/Layr-Labs/eigenda/api/clients"
+	contractEigenDAServiceManager "github.com/Layr-Labs/eigenda/contracts/bindings/EigenDAServiceManager"
+	"github.com/Layr-Labs/eigenda/core"
+	"github.com/Layr-Labs/eigenda/encoding"
+	"github.com/Layr-Labs/eigenda/retriever/eth"
+	"github.com/Layr-Labs/eigenda/tools/traffic/config"
+	"github.com/Layr-Labs/eigenda/tools/traffic/metrics"
+	"github.com/Layr-Labs/eigenda/tools/traffic/table"
+	"github.com/Layr-Labs/eigensdk-go/logging"
+	gcommon "github.com/ethereum/go-ethereum/common"
+	"math/big"
+	"sync"
+	"time"
+)
+
+// BlobReader reads blobs from the DA network at a configured rate.
+type BlobReader struct {
+	// The context for the generator. All work should cease when this context is cancelled.
+	ctx *context.Context
+
+	// Tracks the number of active goroutines within the generator.
+	waitGroup *sync.WaitGroup
+
+	// All logs should be written using this logger.
+	logger logging.Logger
+
+	// config contains the configuration for the generator.
+	config *config.WorkerConfig
+
+	retriever   clients.RetrievalClient
+	chainClient eth.ChainClient
+
+	// blobsToRead blobs we are required to read a certain number of times.
+	blobsToRead *table.BlobStore
+
+	// metrics for the blob reader.
+	metrics *blobReaderMetrics
+}
+
+type blobReaderMetrics struct {
+	generatorMetrics           metrics.Metrics
+	fetchBatchHeaderMetric     metrics.LatencyMetric
+	fetchBatchHeaderSuccess    metrics.CountMetric
+	fetchBatchHeaderFailure    metrics.CountMetric
+	readLatencyMetric          metrics.LatencyMetric
+	readSuccessMetric          metrics.CountMetric
+	readFailureMetric          metrics.CountMetric
+	recombinationSuccessMetric metrics.CountMetric
+	recombinationFailureMetric metrics.CountMetric
+	validBlobMetric            metrics.CountMetric
+	invalidBlobMetric          metrics.CountMetric
+	operatorSuccessMetrics     map[core.OperatorID]metrics.CountMetric
+	operatorFailureMetrics     map[core.OperatorID]metrics.CountMetric
+	requiredReadPoolSizeMetric metrics.GaugeMetric
+	optionalReadPoolSizeMetric metrics.GaugeMetric
+}
+
+// NewBlobReader creates a new BlobReader instance.
+func NewBlobReader(
+	ctx *context.Context,
+	waitGroup *sync.WaitGroup,
+	logger logging.Logger,
+	config *config.WorkerConfig,
+	retriever clients.RetrievalClient,
+	chainClient eth.ChainClient,
+	blobStore *table.BlobStore,
+	generatorMetrics metrics.Metrics) BlobReader {
+
+	return BlobReader{
+		ctx:         ctx,
+		waitGroup:   waitGroup,
+		logger:      logger,
+		config:      config,
+		retriever:   retriever,
+		chainClient: chainClient,
+		blobsToRead: blobStore,
+		metrics: &blobReaderMetrics{
+			generatorMetrics:           generatorMetrics,
+			fetchBatchHeaderMetric:     generatorMetrics.NewLatencyMetric("fetch_batch_header"),
+			fetchBatchHeaderSuccess:    generatorMetrics.NewCountMetric("fetch_batch_header_success"),
```

Comment:
How does it compare to having a metric with a "status" label for success/failure or valid/invalid (instead of one metric for each status)?

Reply:
The way this code is currently set up, all of the latency metrics use the same base metric name but with different labels. The same is true for the count metrics and the gauge metrics. If anything, I feel like I'm using labels too much right now. 🙃 

This metrics pattern is something that extends to all three components in the traffic pattern (the writer, the status checker, and the reader). Should I work on this as a part of this PR, or split that work into a separate one?

Reply:
sounds like a plan

Reply:
Makes sense. Will address this in the follow up PR.